lib/bolt/error.rb

```diff
@@ -39,6 +39,14 @@ def self.unknown_task(task)
     def self.unknown_plan(plan)
       "Could not find a plan named \"#{plan}\". For a list of available plans, run \"bolt plan show\""
     end
+
+    def self.to_puppet_issue(issue, action)
+      Puppet::ParseErrorWithIssue.from_issue_and_stack(
```

Comment:
The fact that the error message is hardcoded here is a bit strange given this has a really generic method name.

It might make more sense to define this issue somewhere in `Bolt::PAL` and raise it directly.

```rb
module Bolt
  class PAL
    PLAN_OPERATION_NOT_SUPPORTED_WHEN_COMPILING = Puppet::Pops::Issues.issue :PLAN_OPERATION_NOT_SUPPORTED_WHEN_COMPILING, :action do
      "Plan language function '#{action}' cannot be used from declarative manifest code or apply blocks"
    end
  end
end
```

and then raise it like

```rb
  raise Puppet::ParseErrorWithIssue.from_issue_and_stack(:PLAN_OPERATION_NOT_SUPPORTED_WHEN_COMPILING, action: 'add_facts')
```

Reply:
You are right. I followed the pattern of extending puppet as we do with logging by organizing the issues into a `Bolt::PAL::Issues` module stored in `bolt/pal/issues` which is loaded in `PAL.load_puppet`.